File: app/src/main/java/com/kunkunapp/allvideodowloader/utils/HistorySQLite.java
package com.kunkunapp.allvideodowloader.utils;

import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class HistorySQLite extends SQLiteOpenHelper {
    private static final String VISITED_PAGES = "visited_pages";
    private static final String TITLE = "title";
    private SQLiteDatabase dB;

    public HistorySQLite(Context context) {
        super(context, "history.db", null, 2);
        dB = getWritableDatabase();
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL("CREATE TABLE visited_pages (title TEXT, link TEXT, time TEXT)");
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        if (newVersion > oldVersion) {
            Log.v("Database Upgrade", "Database version higher than old.");
            clearHistory();
        }
    }

    public void addPageToHistory(VisitedPage page) {
        ContentValues v = new ContentValues();
        v.put(TITLE, page.title);
        v.put("link", page.link);
        Date time = Calendar.getInstance().getTime();
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy MM dd HH mm ss SSS",
                Locale.getDefault());
        v.put("time", simpleDateFormat.format(time));
        if (dB.update(VISITED_PAGES, v, "link = '" + page.link + "'", null) <= 0) {
            dB.insert(VISITED_PAGES, null, v);
        }
    }

    public void deleteFromHistory(String link) {
        dB.delete(VISITED_PAGES, "link = '" + link + "'", null);
    }

    public void clearHistory() {
        dB.execSQL("DELETE FROM visited_pages");
    }

    @SuppressLint("Range")
    public List<VisitedPage> getAllVisitedPages() {
        Cursor c = dB.query(VISITED_PAGES, new String[]{TITLE, "link"}, null, null, null,
                null, "time DESC");
        List<VisitedPage> pages = new ArrayList<>();
        while (c.moveToNext()) {
            VisitedPage page = new VisitedPage();
            page.title = c.getString(c.getColumnIndex(TITLE));
            page.link = c.getString(c.getColumnIndex("link"));
            pages.add(page);
        }
        c.close();
        return pages;
    }

    @SuppressLint("Range")
    public List<VisitedPage> getVisitedPagesByKeyword(String keyword) {
        Cursor c = dB.query(VISITED_PAGES, new String[]{TITLE, "link"}, "title LIKE '%" +
                keyword + "%'", null, null, null, "time DESC");
        List<VisitedPage> pages = new ArrayList<>();
        while (c.moveToNext()) {
            VisitedPage page = new VisitedPage();
            page.title = c.getString(c.getColumnIndex(TITLE));
            page.link = c.getString(c.getColumnIndex("link"));
            pages.add(page);
        }
        c.close();
        return pages;
    }
}

File: app/src/main/java/com/cyberIyke/allvideodowloader/utils/HistorySQLite.kt
package com.cyberIyke.allvideodowloader.utils

import android.annotation.SuppressLint
import android.content.*
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import com.cyberIyke.allvideodowloader.utils.VisitedPage
import java.text.SimpleDateFormat
import java.util.*

class HistorySQLite(context: Context?) : SQLiteOpenHelper(context, "history.db", null, 2) {
    private val dB: SQLiteDatabase

    init {
        dB = writableDatabase
    }

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL("CREATE TABLE visited_pages (title TEXT, link TEXT, time TEXT)")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (newVersion > oldVersion) {
            Log.v("Database Upgrade", "Database version higher than old.")
            clearHistory()
        }
    }

    fun addPageToHistory(page: VisitedPage) {
        val v = ContentValues()
        v.put(TITLE, page.title)
        v.put("link", page.link)
        val time = Calendar.getInstance().time
        val simpleDateFormat = SimpleDateFormat(
            "yyyy MM dd HH mm ss SSS",
            Locale.getDefault()
        )
        v.put("time", simpleDateFormat.format(time))
        if (dB.update(VISITED_PAGES, v, "link = '" + page.link + "'", null) <= 0) {
            dB.insert(VISITED_PAGES, null, v)
        }
    }

    fun deleteFromHistory(link: String) {
        dB.delete(VISITED_PAGES, "link = '$link'", null)
    }

    fun clearHistory() {
        dB.execSQL("DELETE FROM visited_pages")
    }

    @get:SuppressLint("Range")
    val allVisitedPages: MutableList<VisitedPage>
        get() {
            val c = dB.query(
                VISITED_PAGES, arrayOf(TITLE, "link"), null, null, null,
                null, "time DESC"
            )
            val pages: MutableList<VisitedPage> = ArrayList()
            while (c.moveToNext()) {
                val page = VisitedPage()
                page.title = c.getString(c.getColumnIndex(TITLE))
                page.link = c.getString(c.getColumnIndex("link"))
                pages.add(page)
            }
            c.close()
            return pages
        }

    @SuppressLint("Range")
    fun getVisitedPagesByKeyword(keyword: String): List<VisitedPage> {
        val c = dB.query(
            VISITED_PAGES, arrayOf(TITLE, "link"), "title LIKE '%" +
                    keyword + "%'", null, null, null, "time DESC"
        )
        val pages: MutableList<VisitedPage> = ArrayList()
        while (c.moveToNext()) {
            val page = VisitedPage()
            page.title = c.getString(c.getColumnIndex(TITLE))
            page.link = c.getString(c.getColumnIndex("link"))
            pages.add(page)
        }
        c.close()
        return pages
    }

    companion object {
        private const val VISITED_PAGES = "visited_pages"
        private const val TITLE = "title"
    }
}